### PR TITLE
Itsmekori/eng 340 create node scripts to update puzzles in database (DO NOT MERGE)

### PR DESCRIPTION
**Pull request by IvySwing**

### CODE REVIEW

File: scripts/import.mjs [hunk 1]
```diff
@@ -0,0 +1,153 @@
+import fs from "fs";
+import { URL } from "url";
+import fetch from "node-fetch";
+import env from "@next/env";
+
+// Set admin secret to process.env.HASURA_GRAPHQL_ADMIN_SECRET
+env.loadEnvConfig(process.cwd());
+
+const gql = `query MyQuery($siteName: String) {
```
Comment (illepic) [inline scripts/import.mjs]: Give it a name.

File: scripts/import.mjs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  puzzles(where: {simple_name: {_eq: $siteName}}) {
+    final_step
+    input_type
+    fail_route
+    fail_message
+  }
+}`;
+
+// mutation to update puzzle information
+const gqlModify = `mutation MyMutation($simple_name: String, $input_type: puzzle_input_type_enum = boxes, $challenge: String = "", $fail_message: String = "", $landing_message: String = "", $solution: String = "", $success_message: String = "") {
```
Comment (illepic) [inline scripts/import.mjs]: Mutation with object instead is an option.

File: scripts/import.mjs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      input_type
+      simple_name
+    }
+  }
+}
+
+`;
+
+// query to retreive current puzzle information
+const gqlAllPuzzles = `query MyPuzzles {
```
Comment (illepic) [inline scripts/import.mjs]: Give name.

File: scripts/import.mjs (hunk 2 above)
Comment (illepic) [inline scripts/import.mjs]: Give name.

File: tsconfig.json [hunk 4]
```diff
@@ -26,5 +26,11 @@
...
   },
   "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],
-  "exclude": ["node_modules"]
+  "exclude": ["node_modules"],
+  "ts-node": {
+    "require": ["tsconfig-paths/register"],
+    "compilerOptions": {
+      "module": "commonjs"
+    }
+  }
```
Comment (illepic) [inline tsconfig.json]: Nuke this since not needed.

File: package.json [hunk 5]
```diff
@@ -78,6 +78,7 @@
     "prettier": "^2.7.1",
     "start-server-and-test": "^1.14.0",
     "tailwindcss": "^3.1.8",
+    "tsconfig-paths": "^4.1.0",
```
Comment (illepic) [inline package.json]: Nuke this since not needed.

File: scripts/import.mjs (hunk 1 above)
Comment (illepic) [inline scripts/import.mjs]: Nuke this one, dupe'd below.

File: scripts/import.mjs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+      "x-hasura-admin-secret": process.env.HASURA_GRAPHQL_ADMIN_SECRET,
+    },
+    body: JSON.stringify({
+      operationName: "MyQuery",
+      query: gql,
+      variables: { siteName: "notright" },
+    }),
+  });
+  const results = await res.json();
+  console.log(results);
```
Comment (illepic) [inline scripts/import.mjs]: Nuke.

File: scripts/import.mjs [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+  // Example: read json
+  fs.readFile(
+    new URL("./csvjson.json", import.meta.url).pathname,
+    (_err, data) => {
+      const changes = JSON.parse(data);
+      const puzzles = changes.map((change) => String(change.simple_name));
+      console.log(puzzles);
+    }
+  );
```
Comment (illepic) [inline scripts/import.mjs]: Nuke.

File: scripts/import.mjs [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+        "content-type": "application/json",
+        "x-hasura-admin-secret": process.env.HASURA_GRAPHQL_ADMIN_SECRET,
+      },
+      body: JSON.stringify({
+        operationName: "MyPuzzles",
+        query: gqlAllPuzzles,
+      }),
+    }
+  );
+  const getResults = await getPuzzles.json();
```
Comment (illepic) [inline scripts/import.mjs]: Variables that start with a verb are always a function! Please change var name to not start with verb, ie call this `ogPuzzles`.

File: scripts/import.mjs [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    return String(newData) ? String(newData) : oldData;
+  };
+
+  fs.readFile(
+    new URL("./updatedcsvjson.json", import.meta.url).pathname,
+    (_err, data) => {
+      const changes = JSON.parse(data);
+      changes.forEach((change) => {
+        const oldPuzzle = getResults.data.puzzles.find(
+          (puzzle) => puzzle.simple_name === String(change.simple_name)
```
Comment (illepic) [inline scripts/import.mjs:74]: Remove `String()` wherever possible

File: scripts/import.mjs [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        );
+
+        updatePuzzles({
+          simple_name: String(change.simple_name),
+          challenge: String(change.Challenge),
+          input_type: "boxes",
+          landing_message: String(change["Instructions (old landing_message)"]),
+          solution: getGoodData(change["Input (Passcode)"], oldPuzzle.solution),
```
Comment (illepic) [inline scripts/import.mjs:82]:
```
solution: !change["Instructions (old landing_message)"] ? oldPuzzle.solution : change["Instructions (old landing_message)"]
```

File: scripts/import.mjs [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+  //       query: gqlModify,
+  //       variables: { simple_name: "communitycode", input_type: "boxes" },
+  //     }),
+  //   });
+  //   const testResults = await testRes.json();
+  //   console.log("test results", testResults.data.update_puzzles.returning);
+
+  const getGoodData = (newData, oldData) => {
+    return String(newData) ? String(newData) : oldData;
+  };
```
Comment (illepic) [inline scripts/import.mjs]:
```
const getGoodData = (newData, oldData) => newData || oldData;
```

File: scripts/import.mjs [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    new URL("./csvjson.json", import.meta.url).pathname,
+    (_err, data) => {
+      const changes = JSON.parse(data);
+      const puzzles = changes.map((change) => String(change.simple_name));
+      console.log(puzzles);
+    }
+  );
+
+  // function to update by simple name
+  const updatePuzzles = async (variables) => {
```
Comment (illepic) [inline scripts/import.mjs]: Change `variables` to something relevant.

**Review by illepic — CHANGES_REQUESTED** [review]

File: scripts/import.mjs [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    // const testRes = await fetch("https://hasura-2v34.onrender.com/v1/graphql", {
+    //   method: "POST",
+    //   headers: {
+    //     "content-type": "application/json",
+    //     "x-hasura-admin-secret": process.env.HASURA_GRAPHQL_ADMIN_SECRET,
+    //   },
+    //   body: JSON.stringify({
+    //     operationName: "changePuzzle",
+    //     query: gqlModify,
+    //     newPuzzInfo,
```
Comment (stormcloud266) [inline scripts/import.mjs]: this will need to be `variables: newPuzzInfo`

File: yarn.lock [hunk 14]
```diff
@@ -9415,6 +9415,15 @@ tsconfig-paths@^3.14.1:
...
 
+tsconfig-paths@^4.1.0:
+  version "4.1.0"
+  resolved "https://registry.yarnpkg.com/tsconfig-paths/-/tsconfig-paths-4.1.0.tgz#f8ef7d467f08ae3a695335bf1ece088c5538d2c1"
+  integrity sha512-AHx4Euop/dXFC+Vx589alFba8QItjF+8hf8LtmuiCwHyI4rHXQtOOENaM8kvYf5fR0dRChy3wzWIZ9WbB7FWow==
+  dependencies:
+    json5 "^2.2.1"
+    minimist "^1.2.6"
+    strip-bom "^3.0.0"
+
```
Comment (stormcloud266) [inline yarn.lock]: If `tsconfig-paths` was removed from the package.json, let's also run `yarn remove tsconfig-paths` so it deletes it from here as well.

**Review by stormcloud266 — CHANGES_REQUESTED** [review]

**Review by stormcloud266 — APPROVED** [review]
Great job this rocks

**Review by illepic — APPROVED** [review]

**Pull request closed without merging**